File: src/macroaicomp/models/arms_race.py
"""Richardson's arms race model.

Reference: Bongers, A., Gómez, T. and Torres, J.L. (2019), "An Introduction
to Computational Macroeconomics", Chapter 1. Vernon Press.
"""

from dataclasses import dataclass

import numpy as np


@dataclass
class ArmsRaceParams:
    """Calibración del modelo de carrera armamentística de Richardson (eq. 1.9-1.11).

    Parámetros
    ----------
    alpha : float
        Elasticidad de dx1 respecto a x1 (coeficiente de fatiga/gasto).
    beta : float
        Elasticidad de dx1 respecto a x2 (coeficiente de defensa/reacción).
    gamma : float
        Elasticidad de dx2 respecto a x1 (coeficiente de defensa/reacción).
    delta : float
        Elasticidad de dx2 respecto a x2 (coeficiente de fatiga/gasto).
    theta : float
        Elasticidad de dx1 respecto a z1 (coeficiente de agravio/hostilidad).
    eta : float
        Elasticidad de dx2 respecto a z2 (coeficiente de agravio/hostilidad).
    """

    alpha: float
    beta: float
    gamma: float
    delta: float
    theta: float
    eta: float


def coefficient_matrices(params: ArmsRaceParams) -> tuple[np.ndarray, np.ndarray]:
    """Construye las matrices de coeficientes A y B del sistema dinámico (eq. 1.7-1.8).

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo.

    Returns
    -------
    tuple[np.ndarray, np.ndarray]
        Matriz A (2x2, coeficientes de las variables endógenas) y
        matriz B (2x2, coeficientes de las variables exógenas).
    """
    a = np.array([[-params.alpha, params.beta], [params.gamma, -params.delta]])
    b = np.array([[params.theta, 0.0], [0.0, params.eta]])
    return a, b


def steady_state(params: ArmsRaceParams, z: np.ndarray) -> np.ndarray:
    """Calcula el estado estacionario del sistema (eq. 1.14).

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo.
    z : np.ndarray
        Variables exógenas (z1, z2).

    Returns
    -------
    np.ndarray
        Valores de estado estacionario (x1_bar, x2_bar).
    """
    a, b = coefficient_matrices(params)
    return -np.linalg.inv(a) @ b @ z


def eigenvalues(params: ArmsRaceParams) -> np.ndarray:
    """Calcula los autovalores de la matriz A (eq. 1.20-1.25).

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo.

    Returns
    -------
    np.ndarray
        Autovalores (lambda1, lambda2) de la matriz A.
    """
    a, _ = coefficient_matrices(params)
    return np.linalg.eigvals(a)


def is_saddle_path(params: ArmsRaceParams) -> bool:
    """Clasifica el estado estacionario como un punto de silla o globalmente estable.

    Siguiendo el criterio de estabilidad del Apéndice A, un autovalor real
    lambda es estable si |lambda + 1| < 1. El estado estacionario es un punto
    de silla si exactamente uno de los dos autovalores es estable.

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo.

    Returns
    -------
    bool
        True si el estado estacionario es un punto de silla, False en caso contrario.
    """
    moduli = np.abs(eigenvalues(params) + 1)
    return bool((moduli < 1).sum() == 1)


def simulate(
    params: ArmsRaceParams,
    z_initial: np.ndarray,
    z_final: np.ndarray,
    periods: int = 30,
    shock_period: int = 1,
) -> tuple[np.ndarray, np.ndarray]:
    """Simula la dinámica de transición tras un shock (Sección 1.5).

    Reproduce la recursión de la hoja de cálculo "ICM-1-1.xls":
    comenzando desde el estado estacionario inicial, x_(t+1) = x_t + dx_t, donde
    dx_t se evalúa con z_initial para t < shock_period y con z_final
    a partir de shock_period en adelante. Válido para calibraciones globalmente
    estables.

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo.
    z_initial : np.ndarray
        Variables exógenas antes del shock.
    z_final : np.ndarray
        Variables exógenas a partir del shock.
    periods : int
        Número de períodos a simular (el libro usa 30).
    shock_period : int
        Período en el que z cambia de z_initial a z_final.

    Returns
    -------
    tuple[np.ndarray, np.ndarray]
        Trayectorias temporales (x1, x2), ambas de longitud `periods`.
    """
    a, b = coefficient_matrices(params)
    x = np.zeros((periods, 2))
    x[0] = steady_state(params, z_initial)
    for t in range(periods - 1):
        z_t = z_final if t + 1 >= shock_period else z_initial
        # Ecuación de recurrencia exacta periodo a periodo:
        x[t + 1] = x[t] + (a @ x[t] + b @ z_t)
    return x[:, 0], x[:, 1]


def simulate_saddle_path(
    params: ArmsRaceParams,
    z_initial: np.ndarray,
    z_final: np.ndarray,
    periods: int = 30,
    shock_period: int = 1,
    jump_variable: int = 0,
) -> tuple[np.ndarray, np.ndarray]:
    """Simula la dinámica de transición cuando el estado estacionario es un punto de silla.

    Reproduce la hoja de cálculo "ICM-1-2.xls": la variable de expectativa
    (jump_variable) se reajusta instantáneamente en el periodo del shock para
    situarse sobre la senda estable, mientras que la otra variable continúa su
    dinámica histórica en ese periodo.

    Parameters
    ----------
    params : ArmsRaceParams
        Calibración del modelo (debe producir un punto de silla).
    z_initial : np.ndarray
        Variables exógenas antes del shock.
    z_final : np.ndarray
        Variables exógenas a partir del shock.
    periods : int
        Número de períodos a simular.
    shock_period : int
        Período en el que ocurre la perturbación.
    jump_variable : int
        Índice (0 o 1) de la variable flexible que salta.

    Returns
    -------
    tuple[np.ndarray, np.ndarray]
        Trayectorias temporales (x1, x2), ambas de longitud `periods`.
    """
    if not is_saddle_path(params):
        raise ValueError(
            "La calibración no produce un estado estacionario de punto de silla."
        )

    a, b = coefficient_matrices(params)
    lambdas = eigenvalues(params)
    # Encontramos el autovalor estable (el más cercano al límite de estabilidad)
    stable_lambda = lambdas[np.argmin(np.abs(lambdas + 1))]
    other_variable = 1 - jump_variable
    x_bar_final = steady_state(params, z_final)

    x = np.zeros((periods, 2))
    x[0] = steady_state(params, z_initial)
    for t in range(periods - 1):
        if t + 1 == shock_period:
            # 1. La variable predeterminada evoluciona hacia atrás con las variables previas
            x[t + 1, other_variable] = x[t, other_variable] + (
                a[other_variable] @ x[t] + b[other_variable] @ z_initial
            )
            # 2. La variable flexible (jump) da un salto inmediato para caer en la senda estable (saddle path)
            row = a[jump_variable]
            x[t + 1, jump_variable] = (
                row[other_variable] * x[t + 1, other_variable]
                + b[jump_variable, jump_variable] * z_final[jump_variable]
                + stable_lambda * x_bar_final[jump_variable]
            ) / (stable_lambda - row[jump_variable])
        else:
            # 3. Resto de periodos: evolución normal según el nuevo estado de la economía
            z_t = z_final if t + 1 > shock_period else z_initial
            x[t + 1] = x[t] + (a @ x[t] + b @ z_t)
    return x[:, 0], x[:, 1]

File: src/models/ArmsRace.jl
module ArmsRace

using LinearAlgebra

export ArmsRaceParams,
    coefficient_matrices,
    steady_state,
    eigenvalues,
    is_saddle_path,
    simulate,
    simulate_saddle_path

"""
    ArmsRaceParams

Calibration of Richardson's arms race model (eq. 1.9-1.11).

# Fields
- `alpha::Float64`: Elasticity of dx1 with respect to x1.
- `beta::Float64`: Elasticity of dx1 with respect to x2.
- `gamma::Float64`: Elasticity of dx2 with respect to x1.
- `delta::Float64`: Elasticity of dx2 with respect to x2.
- `theta::Float64`: Elasticity of dx1 with respect to z1.
- `eta::Float64`: Elasticity of dx2 with respect to z2.
"""
struct ArmsRaceParams
    alpha::Float64
    beta::Float64
    gamma::Float64
    delta::Float64
    theta::Float64
    eta::Float64
end

"""
    coefficient_matrices(params::ArmsRaceParams)

Build the A and B matrices of the dynamic system (eq. 1.7-1.8).
"""
function coefficient_matrices(params::ArmsRaceParams)
    A = [-params.alpha params.beta; params.gamma -params.delta]
    B = [params.theta 0.0; 0.0 params.eta]
    return A, B
end

"""
    steady_state(params::ArmsRaceParams, z::AbstractVector)

Compute the steady state of the system (eq. 1.14).
"""
function steady_state(params::ArmsRaceParams, z::AbstractVector)
    A, B = coefficient_matrices(params)
    return -(A \ (B * z))
end

"""
    eigenvalues(params::ArmsRaceParams)

Compute the eigenvalues of matrix A (eq. 1.20-1.25).
"""
function eigenvalues(params::ArmsRaceParams)
    A, _ = coefficient_matrices(params)
    return eigvals(A)
end

"""
    is_saddle_path(params::ArmsRaceParams)

Classify the steady state as a saddle point or globally stable.
"""
function is_saddle_path(params::ArmsRaceParams)
    lambdas = eigenvalues(params)
    moduli = abs.(lambdas .+ 1.0)
    return count(x -> x < 1.0, moduli) == 1
end

"""
    simulate(params::ArmsRaceParams, z_initial::AbstractVector, z_final::AbstractVector, periods::Int=30, shock_period::Int=2)

Simulate the transition dynamics after a shock (Section 1.5).
"""
function simulate(
    params::ArmsRaceParams,
    z_initial::AbstractVector,
    z_final::AbstractVector,
    periods::Int=30,
    shock_period::Int=2,
)
    A, B = coefficient_matrices(params)
    x = zeros(periods, 2)
    x[1, :] = steady_state(params, z_initial)
    for t in 1:(periods - 1)
        z_t = (t + 1 >= shock_period) ? z_final : z_initial
        # x[t+1] = x[t] + A*x[t] + B*z_t
        x[t + 1, :] = x[t, :] + (A * x[t, :] + B * z_t)
    end
    return x[:, 1], x[:, 2]
end

"""
    simulate_saddle_path(params::ArmsRaceParams, z_initial::AbstractVector, z_final::AbstractVector, periods::Int=30, shock_period::Int=2, jump_variable::Int=1)

Simulate transition dynamics when the steady state is a saddle point.
"""
function simulate_saddle_path(
    params::ArmsRaceParams,
    z_initial::AbstractVector,
    z_final::AbstractVector,
    periods::Int=30,
    shock_period::Int=2,
    jump_variable::Int=1,
)
    if !is_saddle_path(params)
        throw(ArgumentError("Calibration does not produce a saddle-point steady state."))
    end

    A, B = coefficient_matrices(params)
    lambdas = eigenvalues(params)
    # Find the eigenvalue closest to stable boundary
    stable_lambda = lambdas[argmin(abs.(lambdas .+ 1.0))]

    other_variable = 3 - jump_variable
    x_bar_final = steady_state(params, z_final)

    x = zeros(periods, 2)
    x[1, :] = steady_state(params, z_initial)

    for t in 1:(periods - 1)
        if t + 1 == shock_period
            # other_variable evolves backward
            x[t + 1, other_variable] =
                x[t, other_variable] +
                A[other_variable, 1] * x[t, 1] +
                A[other_variable, 2] * x[t, 2] +
                B[other_variable, 1] * z_initial[1] +
                B[other_variable, 2] * z_initial[2]

            # jump_variable jumps to saddle path
            row = A[jump_variable, :]
            x[t + 1, jump_variable] =
                (
                    row[other_variable] * x[t + 1, other_variable] +
                    B[jump_variable, jump_variable] * z_final[jump_variable] +
                    stable_lambda * x_bar_final[jump_variable]
                ) / (stable_lambda - row[jump_variable])
        else
            z_t = (t + 1 > shock_period) ? z_final : z_initial
            x[t + 1, :] = x[t, :] + (A * x[t, :] + B * z_t)
        end
    end
    return x[:, 1], x[:, 2]
end

end # module ArmsRace
